File: ios/SmashCore/Sources/SmashCore/Feel/MatchCues.swift
/// A banner the match raises (spec §16.4): its copy and arguments, its style and how long it stands.
public struct Banner: Sendable, Hashable {
    public enum Style: String, Sendable, Hashable, CaseIterable {
        /// A goal of ours, a win: popped big in the sun colour.
        case good
        /// A goal against, a loss, a drill interrupted.
        case bad
        /// A dead-ball reset in a match.
        case warn
        /// The intro, periods, a drill's get ready.
        case info
    }

    public let key: CopyKey
    /// The copy's arguments in order (numbers already as text).
    public let args: [String]
    public let style: Style
    /// Real seconds it stands before it leaves.
    public let seconds: Double

    public init(_ key: CopyKey, _ args: [String] = [], style: Style, seconds: Double) {
        self.key = key
        self.args = args
        self.style = style
        self.seconds = seconds
    }
}

/// A haptic the match plays (spec §8.8); intensities 0–1.
public enum Haptic: Sendable, Hashable {
    /// Light: the aim entering a window, a countdown second.
    case tick(Double)
    /// A body blow: a release, the goal horn.
    case impact(Double)
    /// A sharp transient: a steal, a save, a post.
    case sharp(Double)
}

/// Where a save, a steal or a block pops (§8.8).
public enum PopKind: String, Sendable, Hashable, CaseIterable { case save, steal, block }

/// One reaction to what the match did (spec §8.8, §16.4). `delay` is real seconds from now.
public enum Cue: Sendable, Hashable {
    case banner(Banner)
    /// A sound; one made on the pitch carries the ball's `x` across it (it pans there and follows
    /// §8.6's time scale), every other plays at real speed, centred.
    case sound(SoundCue, x: Double?, delay: Double)
    case haptic(Haptic, delay: Double)
    /// A camera shake of this many metres (the prototype's kick, §8.8).
    case shake(Double)
    case pop(PopKind, x: Double, z: Double)
}

/// What the match's events and frames set off — the banners, sounds, haptics, shakes and pops (spec
/// §8.8, §16.4), decided once for both apps. It keeps the little memory the prototype's messages
/// needed (a face-off after a period's end names the period; a drill's get-ready says why), and
/// the countdown's and the aim's last state. The player's matches get everything; the demo behind
/// the menus (§9) only what is seen — the shakes and pops — never a banner, a sound or a haptic.
public struct MatchCues: Sendable {
    public struct Params: Sendable, Hashable {
        public var banner = Seconds()
        public var shakeGoal = 0.0
        public var shakePost = 0.0
        public var hapticWindow = 0.0
        public var releaseLow = 0.0
        public var releaseHigh = 0.0
        public var releaseSlow = 0.0
        public var releaseFast = 0.0
        public var hapticSharp = 0.0
        public var hapticGoal = 0.0
        public var goalPulses: [Double] = []
        public var hapticAgainst = 0.0
        public var hapticTick = 0.0
        public var shotSlow = 0.0
        public var shotFast = 0.0
        public var countdown = 0
        public var stingDelay = 0.0

        /// The banners' durations (presentation.toml [banner]).
        public struct Seconds: Sendable, Hashable {
            public var versus = 0.0, period = 0.0, periodEnd = 0.0, whistle = 0.0
            public var ready = 0.0, lost = 0.0, goal = 0.0, end = 0.0
            public init() {}
        }

        public init() {}
    }

    private let p: Params
    private let isDrill: Bool
    private let audible: Bool
    private var lastFlow: MatchEvent?
    private var lastTick: Int?
    private var aimWindow: MatchSnapshot.Aim?

    /// `drill`: a drill is on the pitch (§10). `audible`: the player's own match, not the demo.
    public init(_ params: Params, drill: Bool, audible: Bool) {
        p = params
        isDrill = drill
        self.audible = audible
    }

    /// The intro banner of a player's match (not a drill): "Home vs Away".
    public func intro(home: String, away: String) -> [Cue] {
        guard audible, !isDrill else { return [] }
        return [.banner(Banner(.eventVersus, [home, away], style: .info, seconds: p.banner.versus))]
    }

    /// What event `e` sets off; `s` is the match as it stands after the tick that emitted it.
    public mutating func hear(_ e: MatchEvent, _ s: MatchSnapshot) -> [Cue] {
        var out: [Cue] = []
        let b = p.banner
        func banner(_ key: CopyKey, _ args: [String] = [], _ style: Banner.Style, _ seconds: Double) {
            out.append(.banner(Banner(key, args, style: style, seconds: seconds)))
        }
        /// A sound; one made on the pitch (the ball's) is at the ball's x — it pans and follows §8.6.
        func sound(_ cue: SoundCue, onPitch: Bool = false, delay: Double = 0) {
            out.append(.sound(cue, x: onPitch ? s.ball.x : nil, delay: delay))
        }
        func haptic(_ h: Haptic, delay: Double = 0) { out.append(.haptic(h, delay: delay)) }
        func mine(_ i: Int) -> Bool { s.players[i].team == 0 && s.players[i].role != .goalie }

        switch e {
        case .goal(let team, _, _, _):
            out.append(.shake(p.shakeGoal))
            sound(.matchWhistleShort)
            sound(.uiConfettiPop)
            if team == 0 {
                banner(.eventGoal, [], .good, b.goal)
                sound(.matchGoalHorn)
                sound(.matchGoalCheer)
                for t in p.goalPulses { haptic(.impact(p.hapticGoal), delay: t) }
            } else {
                banner(.eventGoalAgainst, [], .bad, b.goal)
                sound(.matchGoalAgainst)
                haptic(.impact(p.hapticAgainst))
            }
        case .post:
            out.append(.shake(p.shakePost))
            sound(.matchPost, onPitch: true)
            haptic(.sharp(p.hapticSharp))
        case .board:
            sound(.matchBoard, onPitch: true)
        case .block(let i):
            sound(.matchBlock, onPitch: true)
            out.append(.pop(.block, x: s.players[i].x, z: s.players[i].z))
        case .save(let i):
            sound(.matchSave, onPitch: true)
            haptic(.sharp(p.hapticSharp))
            out.append(.pop(.save, x: s.players[i].x, z: s.players[i].z))
        case .steal(let by, _):
            sound(.matchSteal, onPitch: true)
            haptic(.sharp(p.hapticSharp))
            out.append(.pop(.steal, x: s.players[by].x, z: s.players[by].z))
        case .pickup(let i):
            if s.players[i].team == 0 { sound(.matchReceive, onPitch: true) }
        case .shot(let by, _):
            let speed = Pitch.length(s.ball.vx, s.ball.vz)
            sound(shot(speed), onPitch: true)
            if mine(by) { haptic(.impact(release(speed))) }
        case .pass(let from, _):
            sound(.matchPass, onPitch: true)
            if mine(from) { haptic(.impact(release(Pitch.length(s.ball.vx, s.ball.vz)))) }
        case .play:
            sound(.matchFaceoffDrop, onPitch: true)
        case .whistle:
            banner(.eventReset, [], .warn, b.whistle)
            sound(.matchWhistleShort)
        case .drillInterrupted(let why):
            let key: CopyKey = switch why {
            case .saved: .eventSaved
            case .stolen: .eventStolen
            case .wrongNet: .eventWrongNet
            case .noAssist: .eventPassFirst
            case .deadBall: .eventReset
            }
            banner(key, [], .bad, b.lost)
            sound(.matchWhistleShort)
        case .ready:
            let key: CopyKey = switch lastFlow {
            case .goal?: .eventNiceAgain
            case .drillInterrupted?: .eventAgain
            default: .eventGetReady
            }
            banner(key, [], .info, b.ready)
        case .periodEnd(let period):
            // The third period's end either leads to overtime or is the end itself (its banner).
            if s.state == .periodEnd {
                if period < Tuning.Match.periods {
                    banner(.eventPeriodEnd, ["\(period)"], .info, b.periodEnd)
                } else {
                    banner(.eventOvertime, [], .info, b.periodEnd)
                }
                sound(.matchWhistleEnd)
            }
        case .faceOff:
            if case .periodEnd? = lastFlow {
                if s.overtime { banner(.eventSuddenDeath, [], .info, b.period) } else { banner(.eventPeriod, ["\(s.period)"], .info, b.period) }
            }
        case .end(let result):
            let key: CopyKey = isDrill ? (result == .won ? .resultDrillWon : .resultTimeUp) : .eventFinal
            banner(key, [], result == .lost ? .bad : .good, b.end)
            sound(.matchWhistleEnd)
            sound(result == .lost ? .matchResultLose : .matchResultWin, delay: p.stingDelay)
        }
        switch e {
        case .goal, .drillInterrupted, .periodEnd, .faceOff, .whistle, .ready, .end: lastFlow = e
        default: break
        }
        return audible ? out : out.filter(\.isSeen)
    }

    /// Once per tick, after its events: the countdown of a period's (or a drill's) last seconds, and
    /// the aim of the player's carrier entering a pass or shot window.
    public mutating func frame(_ s: MatchSnapshot) -> [Cue] {
        guard audible else { return [] }
        var out: [Cue] = []
        if s.state == .play && !s.overtime {
            let secs = Int(s.clock.rounded(.up))
            if secs > p.countdown {
                lastTick = nil
            } else if secs >= 1, secs != lastTick {
                lastTick = secs
                out.append(.sound(.matchCountdownTick, x: nil, delay: 0))
                out.append(.haptic(.tick(p.hapticTick), delay: 0))
            }
        }
        let window: MatchSnapshot.Aim? = s.playerCarrier && s.state == .play ? s.aim.flatMap { $0 == .unassisted ? nil : $0 } : nil
        if let w = window, w != aimWindow { out.append(.haptic(.tick(p.hapticWindow), delay: 0)) }
        aimWindow = window
        return out
    }

    /// A shot's sound by its speed: the slowest third of shotSlow…shotFast soft, the middle medium,
    /// the fastest hard.
    func shot(_ speed: Double) -> SoundCue {
        let third = (p.shotFast - p.shotSlow) / 3
        return speed < p.shotSlow + third ? .matchShotSoft : speed < p.shotSlow + 2 * third ? .matchShotMedium : .matchShotHard
    }

    /// A release's impact: from releaseLow at releaseSlow m/s to releaseHigh at releaseFast.
    func release(_ speed: Double) -> Double { ramp(speed, p.releaseSlow, p.releaseFast, p.releaseLow, p.releaseHigh) }

    func ramp(_ v: Double, _ v0: Double, _ v1: Double, _ out0: Double, _ out1: Double) -> Double {
        guard v1 > v0 else { return out1 }
        let t = min(max((v - v0) / (v1 - v0), 0), 1)
        return out0 + (out1 - out0) * t
    }
}

extension Cue {
    /// A cue the demo keeps: what is seen, not heard or felt.
    var isSeen: Bool {
        switch self {
        case .shake, .pop: true
        case .banner, .sound, .haptic: false
        }
    }
}

File: android/core/src/main/kotlin/in/nann/smashhockey/core/feel/MatchCues.kt
package `in`.nann.smashhockey.core.feel

import `in`.nann.smashhockey.core.generated.CopyKey
import `in`.nann.smashhockey.core.generated.Role
import `in`.nann.smashhockey.core.generated.SoundCue
import `in`.nann.smashhockey.core.generated.Tuning
import `in`.nann.smashhockey.core.match.DrillInterruption
import `in`.nann.smashhockey.core.match.MatchEvent
import `in`.nann.smashhockey.core.match.MatchResult
import `in`.nann.smashhockey.core.match.MatchSnapshot
import `in`.nann.smashhockey.core.match.MatchState
import `in`.nann.smashhockey.core.match.Pitch
import kotlin.math.ceil

/** A banner the match raises (spec §16.4): its copy and arguments, its style and how long it stands. */
data class Banner(val key: CopyKey, val args: List<String> = emptyList(), val style: Style, val seconds: Double) {
    enum class Style {
        /** A goal of ours, a win: popped big in the sun colour. */
        GOOD,
        /** A goal against, a loss, a drill interrupted. */
        BAD,
        /** A dead-ball reset in a match. */
        WARN,
        /** The intro, periods, a drill's get ready. */
        INFO,
    }
}

/** A haptic the match plays (spec §8.8); intensities 0–1. */
sealed interface Haptic {
    /** Light: the aim entering a window, a countdown second. */
    data class Tick(val intensity: Double) : Haptic
    /** A body blow: a release, the goal horn. */
    data class Impact(val intensity: Double) : Haptic
    /** A sharp transient: a steal, a save, a post. */
    data class Sharp(val intensity: Double) : Haptic
}

/** Where a save, a steal or a block pops (§8.8). */
enum class PopKind { SAVE, STEAL, BLOCK }

/** One reaction to what the match did (spec §8.8, §16.4). `delay` is real seconds from now. */
sealed interface Cue {
    data class Show(val banner: Banner) : Cue
    /** A sound; one made on the pitch carries the ball's [x]: it pans there and follows §8.6's time scale. */
    data class Sound(val cue: SoundCue, val x: Double? = null, val delay: Double = 0.0) : Cue
    data class Feel(val haptic: Haptic, val delay: Double = 0.0) : Cue
    /** A camera shake of this many metres (the prototype's kick, §8.8). */
    data class Shake(val metres: Double) : Cue
    data class Pop(val kind: PopKind, val x: Double, val z: Double) : Cue

    /** A cue the demo keeps: what is seen, not heard or felt. */
    val isSeen: Boolean get() = this is Shake || this is Pop
}

/**
 * What the match's events and frames set off — the banners, sounds, haptics, shakes and pops (spec
 * §8.8, §16.4), decided once for both apps (the twin of iOS's MatchCues). It keeps the little memory
 * the prototype's messages needed (a face-off after a period's end names the period; a drill's
 * get-ready says why), and the countdown's and the aim's last state. The player's matches get
 * everything; the demo behind the menus (§9) only what is seen.
 *
 * [drill]: a drill is on the pitch (§10). [audible]: the player's own match, not the demo.
 */
class MatchCues(private val p: Params, private val drill: Boolean, private val audible: Boolean) {
    data class Params(
        val banner: Seconds,
        val shakeGoal: Double,
        val shakePost: Double,
        val hapticWindow: Double,
        val releaseLow: Double,
        val releaseHigh: Double,
        val releaseSlow: Double,
        val releaseFast: Double,
        val hapticSharp: Double,
        val hapticGoal: Double,
        val goalPulses: List<Double>,
        val hapticAgainst: Double,
        val hapticTick: Double,
        val shotSlow: Double,
        val shotFast: Double,
        val countdown: Int,
        val stingDelay: Double,
    ) {
        /** The banners' durations (presentation.toml [banner]). */
        data class Seconds(
            val versus: Double, val period: Double, val periodEnd: Double, val whistle: Double,
            val ready: Double, val lost: Double, val goal: Double, val end: Double,
        )
    }

    private var lastFlow: MatchEvent? = null
    private var lastTick: Int? = null
    private var aimWindow: MatchSnapshot.Aim? = null

    /** The intro banner of a player's match (not a drill): "Home vs Away". */
    fun intro(home: String, away: String): List<Cue> =
        if (!audible || drill) emptyList()
        else listOf(Cue.Show(Banner(CopyKey.EVENT_VERSUS, listOf(home, away), Banner.Style.INFO, p.banner.versus)))

    /** What event [e] sets off; [s] is the match as it stands after the tick that emitted it. */
    fun hear(e: MatchEvent, s: MatchSnapshot): List<Cue> {
        val out = ArrayList<Cue>()
        val b = p.banner
        fun banner(key: CopyKey, args: List<String>, style: Banner.Style, seconds: Double) { out += Cue.Show(Banner(key, args, style, seconds)) }
        /** A sound; one made on the pitch (the ball's) is at the ball's x — it pans and follows §8.6. */
        fun sound(cue: SoundCue, onPitch: Boolean = false, delay: Double = 0.0) { out += Cue.Sound(cue, if (onPitch) s.ball.x else null, delay) }
        fun haptic(h: Haptic, delay: Double = 0.0) { out += Cue.Feel(h, delay) }
        fun mine(i: Int) = s.players[i].team == 0 && s.players[i].role != Role.GOALIE
        val none = emptyList<String>()

        when (e) {
            is MatchEvent.Goal -> {
                out += Cue.Shake(p.shakeGoal)
                sound(SoundCue.MATCH_WHISTLE_SHORT)
                sound(SoundCue.UI_CONFETTI_POP)
                if (e.team == 0) {
                    banner(CopyKey.EVENT_GOAL, none, Banner.Style.GOOD, b.goal)
                    sound(SoundCue.MATCH_GOAL_HORN)
                    sound(SoundCue.MATCH_GOAL_CHEER)
                    for (t in p.goalPulses) haptic(Haptic.Impact(p.hapticGoal), t)
                } else {
                    banner(CopyKey.EVENT_GOAL_AGAINST, none, Banner.Style.BAD, b.goal)
                    sound(SoundCue.MATCH_GOAL_AGAINST)
                    haptic(Haptic.Impact(p.hapticAgainst))
                }
            }
            MatchEvent.Post -> {
                out += Cue.Shake(p.shakePost)
                sound(SoundCue.MATCH_POST, onPitch = true)
                haptic(Haptic.Sharp(p.hapticSharp))
            }
            is MatchEvent.Board -> sound(SoundCue.MATCH_BOARD, onPitch = true)
            is MatchEvent.Block -> {
                sound(SoundCue.MATCH_BLOCK, onPitch = true)
                out += Cue.Pop(PopKind.BLOCK, s.players[e.by].x, s.players[e.by].z)
            }
            is MatchEvent.Save -> {
                sound(SoundCue.MATCH_SAVE, onPitch = true)
                haptic(Haptic.Sharp(p.hapticSharp))
                out += Cue.Pop(PopKind.SAVE, s.players[e.by].x, s.players[e.by].z)
            }
            is MatchEvent.Steal -> {
                sound(SoundCue.MATCH_STEAL, onPitch = true)
                haptic(Haptic.Sharp(p.hapticSharp))
                out += Cue.Pop(PopKind.STEAL, s.players[e.by].x, s.players[e.by].z)
            }
            is MatchEvent.Pickup -> if (s.players[e.player].team == 0) sound(SoundCue.MATCH_RECEIVE, onPitch = true)
            is MatchEvent.Shot -> {
                val speed = Pitch.length(s.ball.vx, s.ball.vz)
                sound(shot(speed), onPitch = true)
                if (mine(e.by)) haptic(Haptic.Impact(release(speed)))
            }
            is MatchEvent.Pass -> {
                sound(SoundCue.MATCH_PASS, onPitch = true)
                if (mine(e.from)) haptic(Haptic.Impact(release(Pitch.length(s.ball.vx, s.ball.vz))))
            }
            MatchEvent.Play -> sound(SoundCue.MATCH_FACEOFF_DROP, onPitch = true)
            MatchEvent.Whistle -> {
                banner(CopyKey.EVENT_RESET, none, Banner.Style.WARN, b.whistle)
                sound(SoundCue.MATCH_WHISTLE_SHORT)
            }
            is MatchEvent.DrillInterrupted -> {
                val key = when (e.reason) {
                    DrillInterruption.SAVED -> CopyKey.EVENT_SAVED
                    DrillInterruption.STOLEN -> CopyKey.EVENT_STOLEN
                    DrillInterruption.WRONG_NET -> CopyKey.EVENT_WRONG_NET
                    DrillInterruption.NO_ASSIST -> CopyKey.EVENT_PASS_FIRST
                    DrillInterruption.DEAD_BALL -> CopyKey.EVENT_RESET
                }
                banner(key, none, Banner.Style.BAD, b.lost)
                sound(SoundCue.MATCH_WHISTLE_SHORT)
            }
            MatchEvent.Ready -> {
                val key = when (lastFlow) {
                    is MatchEvent.Goal -> CopyKey.EVENT_NICE_AGAIN
                    is MatchEvent.DrillInterrupted -> CopyKey.EVENT_AGAIN
                    else -> CopyKey.EVENT_GET_READY
                }
                banner(key, none, Banner.Style.INFO, b.ready)
            }
            is MatchEvent.PeriodEnd -> {
                // The third period's end either leads to overtime or is the end itself (its banner).
                if (s.state == MatchState.PERIOD_END) {
                    if (e.period < Tuning.Match.periods) banner(CopyKey.EVENT_PERIOD_END, listOf("${e.period}"), Banner.Style.INFO, b.periodEnd)
                    else banner(CopyKey.EVENT_OVERTIME, none, Banner.Style.INFO, b.periodEnd)
                    sound(SoundCue.MATCH_WHISTLE_END)
                }
            }
            is MatchEvent.FaceOff -> if (lastFlow is MatchEvent.PeriodEnd) {
                if (s.overtime) banner(CopyKey.EVENT_SUDDEN_DEATH, none, Banner.Style.INFO, b.period)
                else banner(CopyKey.EVENT_PERIOD, listOf("${s.period}"), Banner.Style.INFO, b.period)
            }
            is MatchEvent.End -> {
                val key = if (drill) (if (e.result == MatchResult.WON) CopyKey.RESULT_DRILL_WON else CopyKey.RESULT_TIME_UP) else CopyKey.EVENT_FINAL
                banner(key, none, if (e.result == MatchResult.LOST) Banner.Style.BAD else Banner.Style.GOOD, b.end)
                sound(SoundCue.MATCH_WHISTLE_END)
                sound(if (e.result == MatchResult.LOST) SoundCue.MATCH_RESULT_LOSE else SoundCue.MATCH_RESULT_WIN, delay = p.stingDelay)
            }
        }
        when (e) {
            is MatchEvent.Goal, is MatchEvent.DrillInterrupted, is MatchEvent.PeriodEnd, is MatchEvent.FaceOff,
            MatchEvent.Whistle, MatchEvent.Ready, is MatchEvent.End -> lastFlow = e
            else -> Unit
        }
        return if (audible) out else out.filter { it.isSeen }
    }

    /**
     * Once per tick, after its events: the countdown of a period's (or a drill's) last seconds, and
     * the aim of the player's carrier entering a pass or shot window.
     */
    fun frame(s: MatchSnapshot): List<Cue> {
        if (!audible) return emptyList()
        val out = ArrayList<Cue>()
        if (s.state == MatchState.PLAY && !s.overtime) {
            val secs = ceil(s.clock).toInt()
            if (secs > p.countdown) {
                lastTick = null
            } else if (secs >= 1 && secs != lastTick) {
                lastTick = secs
                out += Cue.Sound(SoundCue.MATCH_COUNTDOWN_TICK)
                out += Cue.Feel(Haptic.Tick(p.hapticTick))
            }
        }
        val window = if (s.playerCarrier && s.state == MatchState.PLAY) s.aim?.takeIf { it != MatchSnapshot.Aim.Unassisted } else null
        if (window != null && window != aimWindow) out += Cue.Feel(Haptic.Tick(p.hapticWindow))
        aimWindow = window
        return out
    }

    /** A shot's sound by its speed: the slowest third of shotSlow…shotFast soft, the middle medium, the fastest hard. */
    internal fun shot(speed: Double): SoundCue {
        val third = (p.shotFast - p.shotSlow) / 3
        return when {
            speed < p.shotSlow + third -> SoundCue.MATCH_SHOT_SOFT
            speed < p.shotSlow + 2 * third -> SoundCue.MATCH_SHOT_MEDIUM
            else -> SoundCue.MATCH_SHOT_HARD
        }
    }

    /** A release's impact: from releaseLow at releaseSlow m/s to releaseHigh at releaseFast. */
    internal fun release(speed: Double) = ramp(speed, p.releaseSlow, p.releaseFast, p.releaseLow, p.releaseHigh)

    internal fun ramp(v: Double, v0: Double, v1: Double, out0: Double, out1: Double): Double {
        if (v1 <= v0) return out1
        val t = ((v - v0) / (v1 - v0)).coerceIn(0.0, 1.0)
        return out0 + (out1 - out0) * t
    }
}
